### PR TITLE
Add hunger/thirst crafting distractions

### PR DESCRIPTION
**Pull request by Arokha**
#### Summary
Interface "Interrupt crafting when starving/parched"

#### Purpose of change
It was mentioned in the discord that it'd be desirable to have crafting interrupted if you become dangerously thirsty or hungry. This implements that in the same way that the 'enemy is very close' interrupts are implemented (which means also that the 'no' button is basically useless but don't blame me for that).

#### Describe the solution
Modified the C++ code in `player_activity::get_distractions()` to add two more distractions. The particular values they use to determine hunger and such are cribbed from `character_oracle.cpp` and are not constants there either.
Also doesn't warn you if the thing you're doing is consuming something since it would be rather superfluous to warn you that you're thirsty while you're drinking.

#### Describe alternatives you've considered
No alternatives per-se, though there could be improvements to this solution, such as adding some sort of hysteresis to `player_activity` to make the 'no' option more useful.

#### Testing
Tested with debug, seems to work fine, you get warnings when your needs meet the appropriate levels specified in the code and you aren't warned when you're drinking things.

### CODE REVIEW

File: src/player_activity.cpp [hunk 1]
```diff
@@ -487,6 +498,22 @@ std::map<distraction_type, std::string> player_activity::get_distractions()
                              g->is_in_dangerous_field()->name() ) );
             }
         }
+        // Nested in the !ACT_AIM to avoid nuisance during combat
+        // If this is too bothersome, maybe a list of just ACT_CRAFT, ACT_DIG etc
+        //if( act_id != ACT_CONSUME && act_id != ACT_EAT_MENU ) {
```
Comment (Night-Pryanik) [inline src/player_activity.cpp:503]:
```suggestion
```

File: src/player_activity.cpp [hunk 2]
```diff
@@ -487,6 +498,22 @@ std::map<distraction_type, std::string> player_activity::get_distractions()
...
+        if( std::find( consuming.begin(), consuming.end(), act_id ) == consuming.end() ) {
+            avatar &player_character = get_avatar();
+            if( !is_distraction_ignored( distraction_type::hunger ) ) {
+                if( player_character.get_hunger() >= 300 && player_character.get_starvation() > 2500 ) {
+                    res.emplace( distraction_type::hunger, "You are at risk of starving!" );
```
Comment (Night-Pryanik) [inline src/player_activity.cpp:508]:
```suggestion
                    res.emplace( distraction_type::hunger, _( "You are at risk of starving!" ) );
```

File: src/player_activity.cpp [hunk 3]
```diff
@@ -487,6 +498,22 @@ std::map<distraction_type, std::string> player_activity::get_distractions()
...
+                }
+            }
+            if( !is_distraction_ignored( distraction_type::thirst ) ) {
+                if( player_character.get_thirst() > 520 ) {
+                    res.emplace( distraction_type::thirst, "You are dangerously dehydrated!" );
```
Comment (Night-Pryanik) [inline src/player_activity.cpp:513]:
```suggestion
                    res.emplace( distraction_type::thirst, _( "You are dangerously dehydrated!" ) );
```

**Comment by mqrause** [comment]
What kind of threshold are these numbers representing and is there a better way to check for that?

**Comment by Arokha** [comment]
> What kind of threshold are these numbers representing and is there a better way to check for that?

The thresholds were taken from character_oracle.cpp as mentioned in the PR and are checked using the same methods as there. They can be adjusted if desirable though.

Thirst is warned at 'parched' level (520 thirst).

Hunger is more complicated and involves some linear interpolation stuff so it's hard to express exactly what '2500' starvation is in comparison to something visible in the game UI. Basically the checks are identical to when an NPC would decide they are starving or dying of thirst since they're from character_oracle, so I thought it was appropriate.

I could alter them to check for the effects you gain when starving instead if that's preferred?

**Comment by mqrause** [comment]
Oops, sorry, apparently I only half read the description.

It'd be better if you used the methods from `character_oracle.cpp` to check the thresholds instead of duplicating the logic here. If the logic (or numbers) there changes for some reason, it's bound to cause issues.

**Comment by Arokha** [comment]
> Oops, sorry, apparently I only half read the description.
> 
> It'd be better if you used the methods from `character_oracle.cpp` to check the thresholds instead of duplicating the logic here. If the logic (or numbers) there changes for some reason, it's bound to cause issues.

Any tips on how to do that? I'm still very new to C++. I dunno how I'd get a reference to an instance of the character oracle class for this particular character (if non-NPCs even have one)

**Comment by mqrause** [comment]
First time looking at it for me, too. The oracle should also work for the player character, but it seems to be meant for ai behavior, so it'd be kinda weird to use it here.

Actually it would be even better to extract the checks into their own `Character` methods. Something like `is_dangerously_dehydrated` or whatever name that does the check and then gets used by both your additions and the oracle.

File: src/player_activity.cpp [hunk 4]
```diff
@@ -34,9 +34,11 @@ static const activity_id ACT_CHOP_LOGS( "ACT_CHOP_LOGS" );
 static const activity_id ACT_CHOP_PLANKS( "ACT_CHOP_PLANKS" );
 static const activity_id ACT_CHOP_TREE( "ACT_CHOP_TREE" );
 static const activity_id ACT_CLEAR_RUBBLE( "ACT_CLEAR_RUBBLE" );
+static const activity_id ACT_CONSUME( "ACT_CONSUME" );
 static const activity_id ACT_CONSUME_DRINK_MENU( "ACT_CONSUME_DRINK_MENU" );
 static const activity_id ACT_CONSUME_FOOD_MENU( "ACT_CONSUME_FOOD_MENU" );
 static const activity_id ACT_CONSUME_MEDS_MENU( "ACT_CONSUME_MEDS_MENU" );
+static const activity_id ACT_CONSUME_FUEL_MENU( "ACT_CONSUME_FUEL_MENU" );
```
Comment (dseguin) [inline src/player_activity.cpp:41]:
```suggestion
static const activity_id ACT_CONSUME_FOOD_MENU( "ACT_CONSUME_FOOD_MENU" );
static const activity_id ACT_CONSUME_FUEL_MENU( "ACT_CONSUME_FUEL_MENU" );
static const activity_id ACT_CONSUME_MEDS_MENU( "ACT_CONSUME_MEDS_MENU" );
```
Just to fix the clang-tidy error about ordered string ids

**Review by dseguin — APPROVED** [review]
The hunger/thirst thresholds are checked by value in other places too, might be worth addressing that separately. Otherwise, no issues with this PR.

File: src/player_activity.cpp [hunk 5]
```diff
@@ -63,6 +65,14 @@ static const activity_id ACT_WORKOUT_HARD( "ACT_WORKOUT_HARD" );
...
 
+static const std::vector<activity_id> consuming {
+    ACT_CONSUME,
+    ACT_EAT_MENU,
+    ACT_CONSUME_FOOD_MENU,
+    ACT_CONSUME_DRINK_MENU,
+    ACT_CONSUME_MEDS_MENU,
+    ACT_CONSUME_FUEL_MENU };
+
 static const efftype_id effect_nausea( "nausea" );
```
Comment (dseguin) [inline src/player_activity.cpp:76]:
```suggestion
static const efftype_id effect_nausea( "nausea" );

static const std::vector<activity_id> consuming {
    ACT_CONSUME,
    ACT_EAT_MENU,
    ACT_CONSUME_FOOD_MENU,
    ACT_CONSUME_DRINK_MENU,
    ACT_CONSUME_MEDS_MENU,
    ACT_CONSUME_FUEL_MENU };
```

**Pull request merged**